jax/interpreters/pxla.py

```diff
@@ -61,144 +61,138 @@
 
 FLAGS = flags.FLAGS
 
-unsafe_map, map = map, safe_map
+unsafe_map, map = map, safe_map  # type: ignore
 
 Index = Union[int, slice, Tuple[Union[int, slice], ...]]
 
 
-def device_put(x, devices: Sequence[xb.xla_client.Device], replicate: bool=False) -> List[xb.xla_client._xla.PyLocalBuffer]:
-  """Call device_put on a sequence of devices and return a flat sequence of buffers."""
-  if replicate:
-    return list(it.chain.from_iterable(xla.device_put(x, device) for device in devices))
-  else:
-    return list(it.chain.from_iterable(xla.device_put(val, device) for val, device in safe_zip(x, devices)))
+class Unstacked(NamedTuple):
+  size: int
+
+class Sharded(NamedTuple):
+  shards: int
 
+"""
+Represents all the ways we can shard a dimension.
+- `None` means no sharding;
+- `Sharded` means that the dimension is split into the specified number of chunks,
+  but the split dimension itself is preserved inside the map;
+- `Unstacked` means that the dimension is split into chunks of size 1, and doesn't
+  appear inside the map.
+"""
+AvalDimSharding = Union[Unstacked, Sharded, None]
 
+class ShardedAxis(NamedTuple):
+  axis: int
+
+class Replicated(NamedTuple):
```

Comment:
I don't think that it would be more accurate. It's `ShardedAxis`, because it refers to an _aval axis_ that's been sharded in the `.sharding` attribute. `Replicated` simply means that it's replicated along a _mesh axis_. Those are two different kinds of axes, so I figured it would be better to say that to each mesh axis you can either assign a sharded axis, or you can replicate the data along that dimension.